flexbe_widget/src/flexbe_widget/behavior_action_server.py

```diff
@@ -102,74 +107,49 @@ def _execute_cb(self, goal):
 			be_selection.behavior_checksum = zlib.adler32(be_content_new)
 
 		# reset state before starting new behavior
-		self._engine_status = None
 		self._current_state = None
 		self._behavior_started = False
 
 		# start new behavior
 		self._pub.publish(be_selection)
 
-		try:
-			rate = rospy.Rate(10)
-			while not rospy.is_shutdown():
-				if self._current_state is not None:
-					self._as.publish_feedback(BehaviorExecutionFeedback(self._current_state))
-					self._current_state = None
-
-				# check if goal has been preempted first
-				if self._as.is_preempt_requested():
-					rospy.loginfo('Behavior execution preempt requested!')
-					self._preempt_pub.publish()
-					rate.sleep()
-					self._as.set_preempted('')
-					break
-
-				if self._engine_status is None:
-					rospy.logdebug_throttle(1, 'No behavior engine status received yet. Waiting for it...')
-					rate.sleep()
-					continue
-
-				if self._engine_status.code == BEStatus.ERROR:
-					rospy.logerr('Failed to run behavior! Check onboard terminal for further infos.')
-					rate.sleep()
-					self._as.set_aborted('')
-					break
-
-				if not self._behavior_started:
-					rospy.logdebug_throttle(1, 'Behavior execution has not yet started. Waiting for it...')
-					rate.sleep()
-					continue
-
-				if self._engine_status.code == BEStatus.FINISHED:
-					result = self._engine_status.args[0] \
-						if len(self._engine_status.args) >= 1 else ''
-					rospy.loginfo('Finished behavior execution with result "%s"!' % result)
-					self._as.set_succeeded(BehaviorExecutionResult(outcome=result))
-					break
-
-				if self._engine_status.code == BEStatus.FAILED:
-					rospy.logerr('Behavior execution failed in state %s!' % str(self._current_state))
-					rate.sleep()
-					self._as.set_aborted('')
-					break
-
-				rate.sleep()
-
-			rospy.loginfo('Ready for next behavior start request.')
-
-		except rospy.ROSInterruptException:
-			pass # allow clean exit on ROS shutdown
 
+	def _preempt_cb(self):
+		self._preempt_pub.publish()
+		rospy.loginfo('Behavior execution preempt requested!')
 
-	def _status_cb(self, msg):
-		self._engine_status = msg
 
-		# check for behavior start here, to avoid race condition between execute_cb and status_cb threads
-		if not self._behavior_started and self._engine_status.code == BEStatus.STARTED:
+	def _status_cb(self, msg):
+		if not self._behavior_started and msg.code == BEStatus.STARTED:
 			self._behavior_started = True
+			self._active_behavior_id = msg.behavior_id
 			rospy.loginfo('Behavior execution has started!')
+		# Ignore status until behavior start was received
+		if not self._behavior_started:
+			return
+
+		if msg.behavior_id != self._active_behavior_id:
+			rospy.logwarn('Ignored status because behavior id differed ({} vs {})!'.format(msg.behavior_id, self._active_behavior_id))
+			return
+
+		if msg.code == BEStatus.ERROR:
```

Comment:
Note: An error can also appear if no behavior is running. In particular, a behavior that fails to start throws an error.